ffi/blitbuffer.lua

```diff
@@ -1826,6 +1826,99 @@ function BB_mt.__index:writePNGFromBGR(filename)
     C.free(cdata)
 end
 
+--jit.opt.start(1)
+--jit.opt.start("loopunroll=10")
+function BB_mt.__index:writeBMP(filename)
+    local function write_uint32(of, data)
+        of:write(string.char(band(data, 255)))
+        data = math.floor(data / 256)
+        of:write(string.char(band(data, 255)))
+        data = math.floor(data / 256)
+        of:write(string.char(band(data, 255)))
+        data = math.floor(data / 256)
+        of:write(string.char(data))
+    end
+
+    local output_channels = 3
+    local w, h = self:getWidth(), self:getHeight()
+    local stride = w * output_channels
+
+    local of, err = io.open(filename, "wb")
+    if err ~= nil then
+        return err
+    end
+
+    local bbdump
+    local data
+    if self:getType() == TYPE_BBRGB24 then
+        data = ffi.cast("unsigned char *", self.data)
+    else
+        bbdump = BB.new(w, h, TYPE_BBRGB24, nil)
+        bbdump:blitFrom(self)
+        data = ffi.cast("unsigned char *", bbdump.data)
+    end
+
+    local filesize = stride * h + 54
+    -- update filesize, if stride is not a multiple of 4
+    if band(stride, 3) ~= 0 then
+        filesize = filesize + (4 - band(stride,3))*h
+    end
+
+    -- bfType (2 Bytes)
+    of:write("BM")
+    -- bfSize (4 Bytes)
+    write_uint32(of, filesize)
+    -- bfReserved (4 Bytes)
+    write_uint32(of, 0)
+    -- bfOffBits (4 Byte)
+    write_uint32(of, 54)
+
+    -- biSize (4 Byte)
+    write_uint32(of, 40)
+    -- biWidth (4 Bytes)
+    write_uint32(of, w)
+     -- biHeight (4 Bytes)
+    write_uint32(of, h)
+    -- biPlanes ( 2 Bytes)
+    of:write(string.char(1), string.char(0))
+    -- biBitCount (2 Bytes)
+    of:write(string.char(output_channels * 8), string.char(0))
+
+    for i = 1, 24 do
+        of:write(string.char(0))
+    end
+
+    local pos
+    -- start with bottom line, because BMP stores from bottom to top
+    for y = h-1, 0, -1 do
+        pos = y * stride
+        for x = 0, w-1 do
+            of:write(string.char(data[pos+2]), string.char(data[pos+1]), string.char(data[pos]))
```

Comment:
Okay, I'd forgotten that this nastiness was BGR, which makes scanline-by-scanline copies a pipedream, so, forget my blitFrom comparisons, it's going to be slower ;).